extras/freenode_taxonomy_reader.rb

```diff
@@ -0,0 +1,59 @@
+require 'socket'
+require 'securerandom'
+require 'timeout'
+
+class FreenodeTaxonomyReader
+  NICKSERV = 'NickServ!NickServ@services.'
+  END_MESSGE_REGEX = /End of (.+) taxonomy|(.+) is not registered/
+
+  def initialize(
+        socket_provider: ->() { ssl_socket },
+        username_provider: ->() { "lobsters-#{SecureRandom.alphanumeric(8)}" }
+      )
+    @socket_provider = socket_provider
+    @nick = username_provider.call
+  end
+
+  def for_user(username)
+    _head, *taxonomy_lines = taxonomy_for(username)
+
+    taxonomy_lines.map do |line|
+      line
+        .split(':')[2,3]
+        .map { |t| t.strip }
+    end.to_h
+  end
+
+  private
+
+  def taxonomy_for(username)
+    taxonomy_lines = []
+    s = @socket_provider.call
+
+    Timeout.timeout(30) do
+      s.write("NICK #{@nick}\r\n")
+      s.write("USER #{@nick} * * :Lobsters\r\n")
+      s.write("PRIVMSG NickServ :TAXONOMY #{username}\r\n")
+
+      while line = s.gets
```

Comment:
Let's see what comes back about format, but if we don't decide to write a bigger parser let's have a max line count of 100 to avoid people stuffing war and peace into their taxonomy and hanging an app worker.

Reply:
While freenode does limit the number of metadata entries per account to something reasonable, that's probably a decent idea nonetheless.